## Упражнения по библиотеке Numpy

In [1]:
import numpy as np

Дан случайный массив:

In [2]:
# create array with random numbers from 0 to 10
arr = 10 * np.random.sample((10,))
arr

array([6.54638139, 1.27003665, 2.48345063, 8.32362704, 5.61358312,
       2.79302274, 4.64734012, 2.52197435, 1.14896786, 0.2018995 ])

**1.** Поменять знак у элементов, значения которых между 3 и 8

In [3]:
arr = np.where((arr > 3) & (arr < 8), -arr, arr)
arr

array([-6.54638139,  1.27003665,  2.48345063,  8.32362704, -5.61358312,
        2.79302274, -4.64734012,  2.52197435,  1.14896786,  0.2018995 ])

**2.** Заменить максимальный элемент случайного массива на 0

In [4]:
arr = np.where(arr == np.amax(arr), 0, arr)
arr

array([-6.54638139,  1.27003665,  2.48345063,  0.        , -5.61358312,
        2.79302274, -4.64734012,  2.52197435,  1.14896786,  0.2018995 ])

Дан двумерный массив

In [5]:
matrix = 10 * np.random.sample((2, 3))
matrix

array([[7.973432  , 5.12300138, 8.62057866],
       [3.8534493 , 3.70361377, 2.99166593]])

**3.** Построить прямое произведение массивов (все комбинации с каждым элементом). На вход подается двумерный массив

In [6]:
res = np.transpose([np.tile(matrix[0], matrix[1].size), np.repeat(matrix[1], matrix[0].size)])
res

array([[7.973432  , 3.8534493 ],
       [5.12300138, 3.8534493 ],
       [8.62057866, 3.8534493 ],
       [7.973432  , 3.70361377],
       [5.12300138, 3.70361377],
       [8.62057866, 3.70361377],
       [7.973432  , 2.99166593],
       [5.12300138, 2.99166593],
       [8.62057866, 2.99166593]])

**4.** Даны 2 массива A (8x3) и B (2x2). Найти строки в A, которые содержат элементы из каждой строки в B, независимо от порядка элементов в B

In [7]:
a = np.random.randint(10, size=(8,3))
b = np.random.randint(10, size=(2, 2))
print(a)
print(b)

c = (a == b[..., np.newaxis, np.newaxis])
num_of_includes_for_each_row = (c.sum(axis=(1, 3)) != 0)

print(a[num_of_includes_for_each_row.all(0)])

[[4 5 9]
 [8 3 6]
 [1 0 4]
 [5 6 8]
 [9 5 8]
 [2 9 3]
 [1 2 7]
 [5 9 3]]
[[4 3]
 [5 6]]
[[4 5 9]
 [8 3 6]
 [5 9 3]]


**5.** Дана 10x3 матрица, найти строки из неравных значений (например строка [2,2,3] остается, строка [3,3,3] удаляется)

In [8]:
matrix = np.random.randint(2, size=(10, 3))
print(f"Before: \n{matrix}")

print(f"After: \n{matrix[(matrix[:,1] != matrix[:,0]) | (matrix[:, 1] != matrix[:, 2]) | (matrix[:, 1] != matrix[:, 2])]}")

Before: 
[[1 0 0]
 [1 0 1]
 [1 0 0]
 [1 0 1]
 [0 1 0]
 [1 0 1]
 [0 0 0]
 [0 1 0]
 [1 1 1]
 [0 0 0]]
After: 
[[1 0 0]
 [1 0 1]
 [1 0 0]
 [1 0 1]
 [0 1 0]
 [1 0 1]
 [0 1 0]]


**6.** Дан двумерный массив. Удалить те строки, которые повторяются

In [9]:
matrix = np.random.randint(2, size=(5, 3))
print(f"Before: \n{matrix}")

matrix = np.unique(matrix, axis=0)
print(f"After: \n{matrix}")

Before: 
[[0 1 0]
 [0 0 0]
 [1 1 0]
 [0 1 0]
 [1 0 0]]
After: 
[[0 0 0]
 [0 1 0]
 [1 0 0]
 [1 1 0]]


______
______

Для каждой из следующих задач (1-5) нужно привести 2 реализации – одна без использования numpy (cчитайте, что там, где на входе или выходе должны быть numpy array, будут просто списки), а вторая полностью векторизованная с использованием numpy (без использования питоновских циклов/map/list comprehension).


__Замечание 1.__ Можно считать, что все указанные объекты непустые (к примеру, в __задаче 1__ на диагонали матрицы есть ненулевые элементы).

__Замечание 2.__ Для большинства задач решение занимает не больше 1-2 строк.

___

* __Задача 1__: Подсчитать произведение ненулевых элементов на диагонали прямоугольной матрицы.  
 Например, для X = np.array([[1, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]]) ответ 3.

In [10]:
x = [[1, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]]

# check size
m = len(x)
n = len(x[0])

mult = 1
for i in range(min(m, n)):
    if x[i][i] != 0:
        mult *= x[i][i]
        
mult

3

In [11]:
x = np.array([[1, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]])

np.prod(np.where(np.diag(x) != 0, np.diag(x), 1))

3

* __Задача 2__: Даны два вектора x и y. Проверить, задают ли они одно и то же мультимножество.  
  Например, для x = np.array([1, 2, 2, 4]), y = np.array([4, 2, 1, 2]) ответ True.

In [12]:
# Complexity: O(n)
import collections

x = [1, 2, 2, 4]
y = [4, 2, 1, 2]

collections.Counter(x) == collections.Counter(y)

True

In [13]:
x = np.array([1, 2, 2, 4])
y = np.array([4, 2, 1, 2])

np.array_equal(np.sort(x), np.sort(y))

True

* __Задача 3__: Найти максимальный элемент в векторе x среди элементов, перед которыми стоит ноль. 
 Например, для x = np.array([6, 2, 0, 3, 0, 0, 5, 7, 0]) ответ 5.

In [14]:
x = [6, 2, 0, 3, 0, 0, 5, 7, 0]

max_value = float('-inf')
for i in range(len(x) - 1):
    if x[i] == 0:
        max_value = max(max_value, x[i + 1])

max_value

5

In [15]:
x = np.array([6, 2, 0, 3, 0, 0, 5, 7, 0])
             
condition = np.where(x[:-1] == 0)
np.max(x[condition[0] + 1])

5

* __Задача 4__: Реализовать кодирование длин серий (Run-length encoding). Для некоторого вектора x необходимо вернуть кортеж из двух векторов одинаковой длины. Первый содержит числа, а второй - сколько раз их нужно повторить.  
 Например, для x = np.array([2, 2, 2, 3, 3, 3, 5]) ответ (np.array([2, 3, 5]), np.array([3, 3, 1])).

In [16]:
import collections

x = [2, 2, 2, 3, 3, 3, 5]

freq = {}
for el in x:
    if el in freq:
        freq[el] += 1
    else:
        freq[el] = 1

(list(freq.keys()), list(freq.values()))


([2, 3, 5], [3, 3, 1])

In [17]:
x = np.array([2, 2, 2, 3, 3, 3, 5])

np.unique(x, return_counts=True)

(array([2, 3, 5]), array([3, 3, 1]))

* __Задача 5__: Даны две выборки объектов - X и Y. Вычислить матрицу евклидовых расстояний между объектами. Сравните с функцией scipy.spatial.distance.cdist по скорости работы.

In [18]:
from scipy.spatial import distance
from math import sqrt

In [32]:
x = [(1, 1), (2, 2), (3, 3)]
y = [(6, 6), (5, 5), (4, 4)]

#distance.cdist(x, y, 'euclidean')

n = len(x)
matrix = [[None] * n for i in range(n)]

for i in range(n):
    for j in range(n):
        matrix[i][j] = sqrt((x[i][0] - y[j][0]) * (x[i][0] - y[j][0]) + (x[i][1] - y[j][1]) * (x[i][1] - y[j][1]))

matrix

[[7.0710678118654755, 5.656854249492381, 4.242640687119285],
 [5.656854249492381, 4.242640687119285, 2.8284271247461903],
 [4.242640687119285, 2.8284271247461903, 1.4142135623730951]]

In [20]:
x = np.array([[1, 1], [2, 2], [3, 3]])
y = np.array([[6, 6], [5, 5], [4, 4]])

#np.linalg.norm(x - y, axis=1)

squares = np.add.outer(np.sum(x * x, axis=1), np.sum(y * y, axis=1))
two_mult = np.dot(x, y.T)

# a^2 + b^2 + c^2 + d^2 - 2(ac + bd) == (a - c)^2 + (b - d)^2
np.sqrt(squares - 2 * two_mult)

array([[7.07106781, 5.65685425, 4.24264069],
       [5.65685425, 4.24264069, 2.82842712],
       [4.24264069, 2.82842712, 1.41421356]])

Let's check time:

In [21]:
from timeit import default_timer as timer

Using **scipy.spatial.distance.cdist**:

In [22]:
start = timer()
distance.cdist(x, y, 'euclidean')
end = timer()

print(f"Time: {end - start}")

Time: 8.689600053912727e-05


Using **NumPy**:

In [23]:
x = np.array([[1, 1], [2, 2], [3, 3]])
y = np.array([[6, 6], [5, 5], [4, 4]])

start = timer()
np.linalg.norm(x - y, axis=1)
end = timer()

print(f"Time: {end - start}")

Time: 0.0001491129996793461


_______
________

* #### __Задача 6__: CrunchieMunchies __*__

Вы работаете в отделе маркетинга пищевой компании MyCrunch, которая разрабатывает новый вид вкусных, полезных злаков под названием **CrunchieMunchies**.

Вы хотите продемонстрировать потребителям, насколько полезны ваши хлопья по сравнению с другими ведущими брендами, поэтому вы собрали данные о питании нескольких разных конкурентов.

Ваша задача - использовать вычисления Numpy для анализа этих данных и доказать, что ваши **СrunchieMunchies** - самый здоровый выбор для потребителей.


In [24]:
import numpy as np

1. Просмотрите файл cereal.csv. Этот файл содержит количества калорий для различных марок хлопьев. Загрузите данные из файла и сохраните их как calorie_stats.

In [25]:
calorie_stats = np.loadtxt("./data/cereal.csv", delimiter=",")
calorie_stats

array([ 70., 120.,  70.,  50., 110., 110., 110., 130.,  90.,  90., 120.,
       110., 120., 110., 110., 110., 100., 110., 110., 110., 100., 110.,
       100., 100., 110., 110., 100., 120., 120., 110., 100., 110., 100.,
       110., 120., 120., 110., 110., 110., 140., 110., 100., 110., 100.,
       150., 150., 160., 100., 120., 140.,  90., 130., 120., 100.,  50.,
        50., 100., 100., 120., 100.,  90., 110., 110.,  80.,  90.,  90.,
       110., 110.,  90., 110., 140., 100., 110., 110., 100., 100., 110.])

2. В одной порции CrunchieMunchies содержится 60 калорий. Насколько выше среднее количество калорий у ваших конкурентов?

Сохраните ответ в переменной average_calories и распечатайте переменную в терминале

In [26]:
average_calories = np.average(calorie_stats)

print(f"Average: {average_calories}")
print(f"Difference: {average_calories - 60}")

Average: 106.88311688311688
Difference: 46.883116883116884


3. Корректно ли среднее количество калорий отражает распределение набора данных? Давайте отсортируем данные и посмотрим.

Отсортируйте данные и сохраните результат в переменной calorie_stats_sorted. Распечатайте отсортированную информацию

In [27]:
calorie_stats_sorted = np.sort(calorie_stats)
calorie_stats_sorted

array([ 50.,  50.,  50.,  70.,  70.,  80.,  90.,  90.,  90.,  90.,  90.,
        90.,  90., 100., 100., 100., 100., 100., 100., 100., 100., 100.,
       100., 100., 100., 100., 100., 100., 100., 100., 110., 110., 110.,
       110., 110., 110., 110., 110., 110., 110., 110., 110., 110., 110.,
       110., 110., 110., 110., 110., 110., 110., 110., 110., 110., 110.,
       110., 110., 110., 110., 120., 120., 120., 120., 120., 120., 120.,
       120., 120., 120., 130., 130., 140., 140., 140., 150., 150., 160.])

4. Похоже, что большинство значений выше среднего. Давайте посмотрим, является ли медиана наиболее корректным показателем набора данных.

Вычислите медиану набора данных и сохраните свой ответ в median_calories. Выведите медиану, чтобы вы могли видеть, как она сравнивается со средним значением.

In [28]:
median_calories = np.median(calorie_stats)
median_calories

110.0

5. В то время как медиана показывает, что по крайней мере половина наших значений составляет более 100 калорий, было бы более впечатляюще показать, что значительная часть конкурентов имеет более высокое количество калорий, чем CrunchieMunchies.

Рассчитайте различные процентили и распечатайте их, пока не найдете наименьший процентиль, превышающий 60 калорий. Сохраните это значение в переменной nth_percentile.

In [29]:
nth_percentile = 0
while np.percentile(calorie_stats, nth_percentile) <= 60:
    nth_percentile += 1

nth_percentile

4

6. Хотя процентиль показывает нам, что у большинства конкурентов количество калорий намного выше, это неудобная концепция для использования в маркетинговых материалах.

Вместо этого давайте подсчитаем процент хлопьев, в которых содержится более 60 калорий на порцию. Сохраните свой ответ в переменной more_calories и распечатайте его

In [30]:
more_calories = calorie_stats[calorie_stats > 60].size / calorie_stats.size
more_calories

0.961038961038961

7. Это действительно высокий процент. Это будет очень полезно, когда мы будем продвигать CrunchieMunchies. Но один вопрос заключается в том, насколько велики различия в наборе данных? Можем ли мы сделать обобщение, что в большинстве злаков содержится около 100 калорий или разброс еще больше?

Рассчитайте величину отклонения, найдя стандартное отклонение, Сохраните свой ответ в calorie_std и распечатайте на терминале. Как мы можем включить эту ценность в наш анализ?

In [31]:
calorie_std = np.std(calorie_stats)
print(calorie_std)

m = calorie_stats.mean()
calorie_stats[abs(calorie_stats  - m) < calorie_std].size / calorie_stats.size

19.35718533390827


0.8181818181818182

8. Напишите короткий абзац, в котором кратко изложите свои выводы и то, как, по вашему мнению, эти данные могут быть использованы в интересах Mycrunch при маркетинге CrunchieMunchies.

Калорийность CrunchieMunchies ниже среднего, что должно порадовать потенциального покупателя. Однако, среднее значение часто не отражает реальной картины из-за возможных выбросов. В таком случае имеет смысл рассмотреть медиану. Заметим, что как минимум половина конкурентов проигрывает CrunchieMunchies, но медиана так же может оказаться довольно спорным показателем в случае выбросов, только уже как в большую, так и в меньшую стороны. Рассчитанный нами процентиль уже можно считать довольно весомым аргументом в пользу нашего продукта, но, как отмечалось выше, это неудобная концепция для использования в маркетинговых материалах. Идеальным фактором для обычного покупателя станет заявление о том, что калорийность CrunchieMunchies меньше, чему у 96% хлопьев других марок.

Отметим так же, что стандартное отклонение невелико, и, как оказалось, среднее значение неплохо отражает реальную картину.